contracts/internal/EverlongPositions.sol

```diff
@@ -0,0 +1,295 @@
+/// @custom:disclaimer The language used in this code is for coding convenience
+///                    only, and is not intended to, and does not, have any
+///                    particular legal or regulatory significance.
+abstract contract EverlongPositions is EverlongBase, IEverlongPositions {
+    using DoubleEndedQueue for DoubleEndedQueue.Bytes32Deque;
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Views                                                   │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @inheritdoc IEverlongPositions
+    function getPositionCount() public view returns (uint256) {
+        return _positions.length();
+    }
+
+    /// @inheritdoc IEverlongPositions
+    function getPosition(
+        uint256 _index
+    ) public view returns (Position memory position) {
+        position = _decodePosition(_positions.at(_index));
+    }
+
+    /// @inheritdoc IEverlongPositions
+    function hasMaturedPositions() public view returns (bool) {
+        // Return false if there are no positions.
+        if (_positions.length() == 0) return false;
+
+        // Return true if the current block timestamp is after
+        // the oldest position's `maturityTime`.
+        return (_decodePosition(_positions.at(0)).maturityTime <=
+            block.timestamp);
+    }
+
+    /// @inheritdoc IEverlongPositions
+    function hasSufficientExcessLiquidity() public view returns (bool) {
+        // Return whether the current excess liquidity is greater than
+        // Hyperdrive's minimum transaction amount.
+        return
+            _excessLiquidity() >=
+            IHyperdrive(_hyperdrive).getPoolConfig().minimumTransactionAmount;
+    }
+
+    // TODO: Consider storing hyperdrive's minimumTransactionAmount.
+    /// @inheritdoc IEverlongPositions
+    function canRebalance() public view returns (bool) {
+        return hasMaturedPositions() || hasSufficientExcessLiquidity();
+    }
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Public                                                  │
+    // ╰─────────────────────────────────────────────────────────╯
+
+    /// @inheritdoc IEverlongPositions
+    function rebalance() public {
+        // Close all mature positions (if present) so that the proceeds can be
+        // used to purchase longs.
+        if (hasMaturedPositions()) {
+            _closeMaturedPositions();
+        }
+
+        // Spend Everlong's excess idle liquidity (if sufficient) on opening a long.
+        if (hasSufficientExcessLiquidity()) {
+            _spendExcessLiquidity();
+        }
+
+        // Emit the `Rebalanced()` event.
+        emit Rebalanced();
+    }
+
+    // ╭─────────────────────────────────────────────────────────╮
+    // │ Virtual                                                 │
+    // ╰─────────────────────────────────────────────────────────╯
```

Comment:
I only included base implementations for now (fully expecting to remove them going forward), can add a TODO for this. 

The reason why I left them virtual is because I saw them as the only things that may change for an Everlong vault based on the underlying Hyperdrive....

`_excessLiquidity()` - may change as we'd want different idle liquidity targets for different hyperdrives (or different everlongs on the same hyperdrive.

`_minOpenLongOutput()/_minOpenLongSharePrice()/_minCloseLongOutput()` - may want tighter slippage guards on everlongs with lower idle liquidity since theyll be rebalancing more often?

It's likely that the strategies used to calculate these values will be the same, in which case Everlongs will likely configure them via some scalar value, and these functions will no longer be virtual. Since I didn't know that forsure yet, I left them as virtual functions.

Reply:
Ok that makes sense. That sounds good to me. I like the idea of adding a todo so that we remember to delete them if we don't use them. Thanks for the comprehensive answer!